crates/runtime/src/datafusion/extension/cache_invalidation.rs

```diff
@@ -0,0 +1,537 @@
+}
+
+impl std::fmt::Debug for CacheInvalidationExec {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("CacheInvalidationExec")
+            .field("table", &self.table)
+            .finish_non_exhaustive()
+    }
+}
+
+impl DisplayAs for CacheInvalidationExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(f, "CacheInvalidationExec: table={}", self.table)
+            }
+            DisplayFormatType::TreeRender => {
+                write!(f, "CacheInvalidationExec")
+            }
+        }
+    }
+}
+
+impl ExecutionPlan for CacheInvalidationExec {
+    fn name(&self) -> &'static str {
+        "CacheInvalidationExec"
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn properties(&self) -> &datafusion::physical_plan::PlanProperties {
+        self.input_exec.properties()
+    }
+
+    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
+        vec![&self.input_exec]
+    }
+
+    fn benefits_from_input_partitioning(&self) -> Vec<bool> {
+        vec![false]
+    }
+
+    fn maintains_input_order(&self) -> Vec<bool> {
+        vec![true; self.children().len()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        if children.len() != 1 {
+            return Err(DataFusionError::Internal(
+                "CacheInvalidationExec requires exactly one child".to_string(),
+            ));
+        }
+
+        let Some(input) = children.into_iter().next() else {
+            unreachable!("should have one input");
+        };
+
+        Ok(Arc::new(Self::new(
+            input,
+            self.table.clone(),
+            Weak::clone(&self.caching),
+        )))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> datafusion::error::Result<SendableRecordBatchStream> {
+        let input_stream = self.input_exec.execute(partition, context)?;
+        let schema = self.input_exec.schema();
+
+        let table = self.table.clone();
+        let caching = Weak::clone(&self.caching);
+
+        // Create an async stream that wraps the input and handles cache invalidation
+        let stream = stream! {
+            let mut input_stream = input_stream;
+            let mut exec_failed = false;
+
+            while let Some(batch_result) = input_stream.next().await {
+                match batch_result {
+                    Ok(batch) => yield Ok(batch),
+                    Err(e) => {
+                        exec_failed = true;
+                        yield Err(e);
+                    }
+                }
+            }
+
+            // If wrapped input_exec completed without error, invalidate the cache
+            if !exec_failed {
```

Comment:
We might improve this further to check for 0 rows modifies / empty result and skip invalidation in this case.